### PR TITLE
Promote develop → main

### LINKED ISSUE #259: Build out test coverage for pure-logic functions (node:test baseline from #253)
## Origin & baseline

#253 added a minimal `node:test` harness — `npm test` runs `node --import tsx --test "src/**/*.test.ts"` (zero extra deps, via the existing `tsx`), tests live beside code as `*.test.ts`, and `*.test.ts` is excluded from the `tsc` build so nothing leaks into `dist/` or npm. The only test so far is `src/config.test.ts` (`parseCommaList`). **This issue builds real coverage on that baseline** — the harness choice is settled; don't re-litigate it.

This is deliberately scoped to **pure-logic functions** (data transforms with no GitHub API calls). That's where bugs like #253 (phantom `""` from a trailing comma) and #257 (foreign repo's board item leaking into a repo-scoped result) actually hide, and they're testable with plain inputs — no mocking. The GitHub API-calling layer is explicitly **out of scope** (see below).

### Tier 1 — high-value, fully pure, easy wins
- `slugify` — `src/github.ts:121` — title → branch slug (lowercasing, punctuation strip, 5-word cap). Also the near-duplicate `slugify` in `src/docs.ts:171` (takes `maxWords`) — worth a test each and a note that two slugifiers exist.
- `bumpVersion` — `src/tools/prepare_release.ts:20` — semver patch/minor/major bump. Classic off-by-one/reset-on-bump territory.
- `isNewer` — `src/server.ts:99` — semver comparison driving the upgrade nudge. Version-compare logic is bug-prone (e.g. `0.10.0` vs `0.9.0`).
- `splitCommitMessage` — `src/tools/commit_and_update.ts:33` — subject/body split.
- `parseRelationships` — `src/github.ts:262` — parses the `## Relationships` section (blocked_by/blocking/parent) from an issue body. Drives the list_issues tree; malformed bodies shouldn't throw.
- `extractBranchFromBody` — `src/github.ts:495` — pulls the **Branch:** link out of an issue body. Load-bearing for create_pull_request/commit_and_update.
- `parseOwnerRepo` — `src/cli/probe.ts:32` **and** `src/github.ts:31` (duplicated) — git remote URL → {owner, repo}. Edge cases: ssh vs https, `.git` suffix, trailing slash. Two copies = worth a shared fixture and a dedup note.
- `renderAutopilotDecisions` — `src/autopilot.ts:29` — decisions array → report block (incl. the empty/null → no-block case from #251's bullet-safety work).
- `priorityRank` — `src/tools/list_issues.ts:25` — priority ordering key.
- `isUntouchedBranch` — `src/tools/close_issue.ts:31` — predicate over commit list.
- `parseCommaList` — `src/config.ts:6` — **done** in #253, listed for completeness.

### Tier 2 — pure string transforms with real edge cases
- `insertChangelogEntry` — `src/docs.ts:122` — inserts an entry under the right `###` heading. (Note: its caller `foldFragmentsIntoChangelog` at `src/docs.ts:208` reads the filesystem — test the pure `insertChangelogEntry` core, not the fs wrapper.)
- `rollChangelogForRelease` — `src/docs.ts:78` — `[Unreleased]` → `## [X.Y.Z]` roll + fresh Unreleased + compare links.
- `getUnreleasedSection` — `src/docs.ts:66`, `getChangeType` — `src/docs.ts:32`, `truncateAtWord` — `src/docs.ts:43`, `buildChangelogEntry` — `src/docs.ts:50`, `buildFragment` — `src/docs.ts:184`.
- `replaceExactly` — `src/tools/prepare_release.ts:31` — exact-count string replace with a labelled throw on miscount (the safety valve for version bumps).
- CLI env round-trip: `parseEnv`/`serializeEnv`/`unquote`/`quoteIfNeeded`/`harvest`/`migrate` — `src/cli/env.ts:53,171,131,140,104,122`. This is the setup-wizard's `.env` read/rewrite logic — quoting, comment preservation, okffs-block boundaries. High edge-case density; good fixture target.
- `renderBanner`/`buildBannerInfo`/`isTrue`/`isOff` — `src/cli/banner.ts`.

### Tier 3 — env-dependent (need a note, not necessarily env mutation)
- `buildBranchName` — `src/github.ts:137` — pure except it reads `config.identifier`. The no-identifier path is directly testable; the identifier path needs env. **General note:** `config` is evaluated from `process.env` once at import into a frozen object, so functions that read `config.*` are awkward to test across env variations. #253 sidestepped this by making `parseCommaList` pure over its input — **prefer that pattern** (take inputs, don't read `config`) for new testable helpers. `parseMergeMethod` (`src/config.ts:22`) reads `process.env[name]` directly and would need env setup.

## Special target — the repo-scoped `projectItems`-filtering shape (from #257)

#257 fixed a class of bug where a shared org board's items leak across repos. Three functions now share a "repo-scoped by construction" shape, but they are **not** identical:
- `getProjectFieldsByIssueNumber` — `src/projects.ts:344` (**#257's fix**) — starts from `repository.issues`, reads each issue's `projectItems`, keeps the one where `project.id` matches. Returns issue# → {status, priority, effort}.
- `getProjectItemForIssue` — `src/projects.ts:311` — single issue, same `projectItems` + `project.id` filter, returns the item id.
- `getOrgIssueFieldValuesByNumber` — `src/projects.ts:403` — repo-scoped **but by a different mechanism**: it reads `issueFieldValues` (org Issue Fields), **not** `projectItems`, and has **no** `project.id` filter. Shares the principle, not the shape.

These are async GraphQL callers, so they're not directly pure. **Recommended approach:** extract the response→map *transform* (given a parsed GraphQL payload, pick the item whose `project.id` matches and key by issue number) into a pure helper, then feed it a **shared fixture** — one mock payload mixing foreign-repo and native project items — asserting foreign items never enter the result. That one fixture covers both `projectItems` functions; `getOrgIssueFieldValuesByNumber` needs its own (different payload shape) but the same "no foreign leakage" assertion. This is exactly the subtle shape bug that's hard to catch by eye and the best argument for fixtures here.

## Out of scope (call out, do not fold in)

- **GitHub API-calling layer** — everything routing through `request`/`graphqlRequest` (most of `src/github.ts`, `src/board.ts`, `src/projects.ts`, `src/issue_types.ts`, and every `src/tools/*.ts` handler, ~40+ functions). Testing these means mocking GitHub responses for real effort and uncertain payoff. **Deferred to a possible separate issue**, not started here. Note (see below) that the current code calls the singleton client directly, so mocking would need `node:test`'s experimental `mock.module` or a refactor to inject the client — decide that when/if that issue is taken up.
- **CI wiring** — gating the npm publish workflow on `npm test` is a **separate decision, deferred**. Not part of this issue.
- **Non-existent-label validation** — the optional idea from #253 (validate a well-formed-but-nonexistent label/assignee before the API call) is a behaviour question, not a test task. Separate issue if pursued.

## node:test fit notes (surfaced now)

- **Pure functions:** perfect fit, no friction — this is why the tranche above is the right start.
- **Import-time env:** `config.ts` reads `process.env` at import into a frozen const. Env-dependent functions (Tier 3) can't easily be re-exercised with different env after import. Mitigation is the #253 pattern (pure over inputs), not env mutation between tests.
- **ESM module mocking:** the API layer calls the singleton `request`/`graphqlRequest` directly (not injected). Mocking them needs `mock.module` (experimental in Node ≥ 22/24) or a dependency-injection refactor. This is the main reason to **not** start with the API layer — flagged now, before a bigger commitment.

## Acceptance

- [ ] Tier 1 pure functions have `*.test.ts` coverage (name → expected, plus malformed/empty inputs that must not throw).
- [ ] The `projectItems` repo-scoping transform is covered by a shared fixture proving no foreign-repo item leaks into a repo-scoped result (the #257 regression).
- [ ] Tier 2 covered as capacity allows; Tier 3 either covered with explicit env setup or documented as env-dependent.
- [ ] Out-of-scope items (API layer, CI, label validation) remain untouched and are left as their own future issues.

**Branch:** `259-okffs-build-out-test-coverage-for`

### LINKED ISSUE #295: update_issue with a body wipes okffs's own metadata (**Branch:** line, ## Relationships), breaking commit_and_update
## Problem

`update_issue` PATCHes `body` straight through to the GitHub REST API with no awareness of the metadata okffs itself writes into the issue body. Any body rewrite therefore silently destroys:

- the `**Branch:**` branch-link line written by `create_issue` (`src/tools/create_issue.ts:124`, `create_issues_from_list.ts:96`, `plan.ts:173`)
- the **`## Relationships`** section written by `link_issues` (`src/tools/link_issues.ts:31-35`)

Losing the **Branch:** line breaks every downstream tool that reads it via `extractBranchFromBody` (`src/github.ts:568`):

- `commit_and_update` → *"Issue #N has no associated branch (no **Branch:** line), so there's nothing to commit to."* (`src/tools/commit_and_update.ts:130`)
- `merge_pull_request` (issue mode) → *"there's no PR to merge"* (`merge_pull_request.ts:109`)
- `create_pull_request` → falls back to inference / requires an explicit `branch`
- `list_issues` / `get_issue` → stop showing the branch

Losing `## Relationships` breaks the parent/child/blocked-by tree in `list_issues`.

Reported from a live repo: an issue's body was updated with `update_issue` after creation, and the next `commit_and_update` failed with "no associated branch". The line had to be restored by hand.

*(Meta: this issue originally quoted the literal branch-line pattern in the first bullet, and `extractBranchFromBody`'s unanchored regex matched that quoted placeholder instead of the real branch line below — so okffs tools showed this issue's branch as `{branch}`. Rewritten to defuse it; the fix should also anchor the extraction to a line start.)*

## Why the current behaviour isn't enough

The tool description does say body "Replaces the existing body — pass the full content, not a fragment", so replace semantics are intentional for *user* content. But the **Branch:** line and `## Relationships` are okffs-owned metadata that the caller never authored and has no reason to know about — the agent rewriting an issue body will essentially never think to re-append them. Replace semantics for user prose shouldn't mean destroying the tool's own bookkeeping.

## Proposed fix

In `update_issue`'s handler, when `body` is provided:

1. Fetch the current issue body.
2. Extract okffs-owned blocks from it — the `**Branch:**` line (reuse `extractBranchFromBody`) and the `## Relationships` section.
3. If the incoming `body` doesn't already contain them, re-append them to the new body before the PATCH.
4. Report it in the tool result (e.g. `body → updated (preserved: **Branch:**, ## Relationships)`), so the behaviour isn't invisible.

Add an explicit escape hatch for a genuine full wipe rather than making preservation unconditional-and-unavoidable — e.g. `preserve_metadata: false` (defaults `true`).

The extraction/merge logic should be a **pure helper** (e.g. extend `src/issue_body.ts`) so it's unit-testable without importing `github.ts`, per the testing convention in CLAUDE.md.

## Also worth considering

`link_issues` appends `## Relationships` with `currentBody.trimEnd() + ...`, which places it *after* the **Branch:** line. That's fine today, but a shared "okffs metadata block" helper used by `create_issue`, `link_issues`, `create_pull_request`'s backfill and `update_issue` would keep ordering consistent and give one place to extend when new metadata is added.

## Acceptance

- `update_issue` with a `body` that omits them leaves the **Branch:** line and `## Relationships` intact on the issue.
- `commit_and_update` still works on that issue immediately afterwards.
- `preserve_metadata: false` reproduces today's raw-replace behaviour.
- Unit tests cover extract + merge (present, absent, already-included, duplicate-avoidance).

**Branch:** `295-okffs-updateissue-with-a-body-wipes`

### LINKED ISSUE #297: Remove the deprecated one-release aliases: hint (commit_and_update), description (create_issue/plan/create_issues_from_list), issue (update_project_status)
## Problem

0.11.0 shipped three param renames, each keeping the old name as a deprecated alias "for one release" (the #279 pattern):

- `commit_and_update`: `hint` → `message` (#290) — alias at `src/tools/commit_and_update.ts` (schema field, both-params error, deprecation warning)
- `create_issue` / `plan` / `create_issues_from_list`: `description` → `body` (#282) — alias fields in each tool schema, resolution in `resolveIssueBody` (`src/issue_body.ts`)
- `update_project_status`: `issue` → `issue_number` (#278) — alias at `src/tools/update_project_status.ts`

The next release is that "one release later", so the aliases come out now per the stated policy.

## Scope

1. Remove the alias fields from the three tool schemas (schemas already reject unknown keys, so a stale caller passing the old name gets an actionable "unknown param" error rather than silent stripping — the #290 behaviour).
2. Simplify `resolveIssueBody` to `body`-only (keep it pure; make a missing body the only error path) and update its unit tests.
3. Remove the both-params error branches and deprecation warnings that only existed for the aliases.
4. Ensure the "unknown key" rejection messages name the canonical param (e.g. passing `hint` should say "use `message`", passing `description` should say "use `body`") so upgrading callers still get pointed the right way.
5. Changelog fragment under `### Removed`.

## Acceptance

- The three tools accept only the canonical params; the old names are rejected with an actionable message naming the canonical replacement.
- `npm test` passes with updated `resolveIssueBody` tests.
- CLAUDE.md's mentions of the deprecated aliases are updated.

**Branch:** `297-okffs-remove-the-deprecated-one-release-aliases`

### LINKED ISSUE #302: Surface promotion-PR review feedback proactively — okffs should detect and drive the Copilot-review fix loop, not wait to be asked
## Problem

The promotion flow requests a Copilot review on the gate PR (`OKFFS_PROMOTION_AUTO_REVIEW` + `OKFFS_PROMOTION_REVIEWERS`), but nothing in okffs ever *checks back* for the resulting feedback. The review lands on GitHub, and in **every repo using okffs** the user has to notice it themselves and prompt the agent ("there are comments on the Copilot review — go fix them") before the `list_pr_review_comments` → fix → `fix_into_base` → reply/resolve loop kicks in. Reported after the okffs 0.12.0 release (PR #300): 4 unresolved Copilot threads sat on the gate PR invisible to the agent workflow.

okffs already owns every piece of the loop (`list_pr_review_comments`, `fix_into_base`, `reply_to_review_comment`, `resolve_review_thread`, the `address_pr_review` prompt) — what's missing is the *trigger*: nothing surfaces "your gate PR has unresolved review threads" inside the okffs conversation.

## Proposed direction

MCP servers can't push notifications, so the lever is enriching the responses of the tools an agent already calls:

1. **`promote_branch` re-runs report review state.** It already reuses an existing open gate PR instead of erroring — on that path, fetch the PR's review threads and include `⚠ N unresolved review thread(s) from <reviewer>` in the response, with an explicit instruction to run the `address_pr_review` loop (fixes via `fix_into_base`, resolve only after the fix PR merges). A re-run of `promote_branch` then becomes the natural "check the release gate" command.

2. **A lightweight gate-status surface.** Either a small `get_promotion_status` tool (open gate PR, mergeable/checks state, unresolved-thread count) or a line in `list_issues` output when an open promotion PR has unresolved threads — so the state is visible at the start of any session, not only when someone thinks to ask.

3. **SERVER_INSTRUCTIONS nudge.** Document the expectation: after `promote_branch` requests a (billable) review, the agent should check back on the gate PR before handing the merge to the user — e.g. by re-calling `promote_branch` or the status tool from (2).

(1) is the highest-value piece and likely sufficient on its own; (2)/(3) are cheap complements. Reviewer feedback on *issue* PRs is out of scope — those already flow through `create_pull_request`/`merge_pull_request`, whose merge gate blocks on unresolved threads and says so.

## Acceptance

- Re-running `promote_branch` while the gate PR has unresolved review threads surfaces the count + reviewer and points the agent at the `address_pr_review` loop (and does NOT re-request the billable review).
- A session that never calls `promote_branch` can still discover the pending feedback (via the chosen (2) surface).
- No behavior change when there are no unresolved threads.

**Branch:** `302-okffs-surface-promotion-pr-review-feedback-proactively`

### PR DESCRIPTION
**Pull request by 2b9sa2owa**
Release 0.12.0 — promotes four merged issues from develop:

**Added** — #302: promotion-PR review feedback is surfaced proactively — `promote_branch` re-runs report the gate PR's unresolved review threads (without re-requesting the billable review), `list_issues` prints a "Release gate" header when threads are pending, and the server instructions tell the agent to check back after an auto-requested review.

**Fixed** — #295: `update_issue` with a `body` no longer wipes okffs-owned issue metadata (the `**Branch:**` line and `## Relationships` section are re-appended when the new body omits them; `preserve_metadata: false` is the raw-replace escape hatch), and `extractBranchFromBody` is anchored to a line start so prose quoting the pattern can't shadow the real branch line.

**Removed** — #297: the three deprecated one-release aliases from 0.11.0 (`hint` → `message` on commit_and_update, `description` → `body` on create_issue/plan/create_issues_from_list, `issue` → `issue_number` on update_project_status); removed names now get an actionable error naming the canonical param. Follow-up #301 (from the Copilot review of this PR) makes `body`/`issue_number` required in the schemas to match.

**Changed** — #259: pure-logic test coverage built out (74 tests) — branch/version/remote/priority helpers, issue-body parsing, commit-message splitting, the CLI `.env` round-trip, the #257 repo-scoped board transforms, and the new review-gate summarization all live in importable pure modules with re-exports keeping importers unchanged.

After merge: tag `v0.12.0` and push it to trigger the CI npm publish.

## Promoting (6 commits)
- Close #295 - update_issue with a body wipes okffs's own metadata (**Branch:** line, ## Relationships), breaking commit_and_update (#296)
- Close #297 - Remove the deprecated one-release aliases: hint (commit_and_update), description (create_issue/plan/create_issues_from_list), issue (update_project_status) (#298)
- Close #259 - Build out test coverage for pure-logic functions (node:test baseline from #253) (#260)
- release: 0.12.0 (#299)
- fix: make body / issue_number required in the tool schemas to match runtime behavior (PR #300 Copilot review) (#301)
- Close #302 - Surface promotion-PR review feedback proactively — okffs should detect and drive the Copilot-review fix loop, not wait to be asked (#303)

### CODE REVIEW

File: src/tools/create_issue.ts [hunk 1]
```diff
@@ -78,9 +78,9 @@ export async function getDescription(): Promise<string> {
 export const inputSchema = z.object({
   title: z.string().describe("Issue title"),
   // `body` is the canonical name (#282), matching update_issue and GitHub's own
-  // field; `description` stays as a deprecated alias for one release (#279 pattern).
+  // field. The `description` alias shipped in 0.11.0 as a one-release
+  // deprecation and was removed in #297.
   body: z.string().optional().describe("Issue body"),
```
[the remark this answers is not in the document]
Comment (2b9sa2owa) [reply src/tools/create_issue.ts]: Fixed in #301 (merged into develop): `body` is now required in create_issue's schema, so a missing body fails validation up front instead of at runtime. `resolveIssueBody`'s error path stays as defense in depth for the shared task-object callers.

File: src/tools/plan.ts [hunk 2]
```diff
@@ -43,11 +43,10 @@ const relationshipSchema = z.object({
 
 const taskSchema = z.object({
   title: z.string().describe("Issue title"),
-  // `body` is the canonical name (#282); `description` is a deprecated alias
-  // for one release (#279 pattern). Note the top-level `description` input (the
-  // free-text work description) is a different field and is unaffected.
+  // `body` is the canonical name (#282); the per-task `description` alias was
+  // removed in #297. Note the top-level `description` input (the free-text work
+  // description) is a different field and is unaffected.
   body: z.string().optional().describe("Issue body"),
```
[the remark this answers is not in the document]
Comment (2b9sa2owa) [reply src/tools/plan.ts]: Fixed in #301 (merged into develop): the per-task `body` is now required in plan's task schema, aligning the declared contract with runtime behavior. (The top-level free-text `description` input is unaffected.)

File: src/tools/create_issues_from_list.ts [hunk 3]
```diff
@@ -20,10 +20,9 @@ export const description =
 
 const taskSchema = z.object({
   title: z.string().describe("Issue title"),
-  // `body` is the canonical name (#282); `description` is a deprecated alias
-  // for one release (#279 pattern).
+  // `body` is the canonical name (#282); the `description` alias was removed
+  // in #297.
   body: z.string().optional().describe("Issue body"),
```
[the remark this answers is not in the document]
Comment (2b9sa2owa) [reply src/tools/create_issues_from_list.ts]: Fixed in #301 (merged into develop): the task schema's `body` is now required, keeping the tool contract consistent with runtime behavior.

File: src/tools/update_project_status.ts [hunk 4]
```diff
@@ -15,10 +15,9 @@ const STATUSES = ["Backlog", "Ready", "In Progress", "Review"] as const;
 
 export const inputSchema = z.object({
   // `issue_number` is the canonical name, matching every other issue-taking
-  // tool (#278). `issue` — this tool's original name — stays as a deprecated
-  // alias for one release so hosts with a cached schema don't hard-break.
+  // tool (#278). The `issue` alias shipped in 0.11.0 as a one-release
+  // deprecation and was removed in #297.
   issue_number: z.number().int().positive().optional().describe("Issue number to move"),
```
[the remark this answers is not in the document]
Comment (2b9sa2owa) [reply src/tools/update_project_status.ts]: Fixed in #301 (merged into develop): `issue_number` is now required in the schema (and the now-unreachable in-handler guard was removed), so the contract is communicated to hosts up front.
